library/Cube/IcingaDb/IcingaDbCube.php

```diff
@@ -0,0 +1,209 @@
+
+namespace Icinga\Module\Cube\IcingaDb;
+
+use Icinga\Module\Cube\Cube;
+use Icinga\Module\Icingadb\Common\Auth;
+use Icinga\Module\Icingadb\Common\Database;
+use ipl\Orm\Query;
+use ipl\Orm\Relation;
+use ipl\Sql\Select;
+
+abstract class IcingaDbCube extends Cube
+{
+    use Auth;
+    use Database;
+
+    /** @var Relation */
+    protected $relation;
+
+    /** @var Query The inner query fetching all required data */
+    protected $innerQuery;
+
+    /** @var Select The rollup query, creating grouped sums over innerQuery */
+    protected $rollupQuery;
+
+    /** @var Select The outer query, orders respecting NULL values, rollup first */
+    protected $fullQuery;
+
+    /** @var boolean */
+    protected $isSliced;
+
+    /** @var array */
+    protected $groupBy;
+
+    /** @var array */
+    protected $columns;
+
+    abstract public function getObjectsFilter();
+
+    public function hasBeenSliced()
+    {
+        return $this->isSliced;
+    }
+
+    public function getRelation()
+    {
+        return $this->relation;
+    }
+
+    /**
+     * An IcingaDbCube must provide a list of all available columns
+     *
+     * This is a key/value array with the key being the fact name / column alias
+     * and
+     *
+     * @return array
+     */
+    abstract public function getAvailableFactColumns();
+
+    /**
+     * @return Query
+     */
+    abstract public function prepareInnerQuery();
+
+    /**
+     * Get our inner query
+     *
+     * Hint: mostly used to get rid of NULL values
+     *
+     * @return Query
+     */
+    public function innerQuery()
+    {
+        if ($this->innerQuery === null) {
+            $this->innerQuery = $this->prepareInnerQuery();
+        }
+
+        return $this->innerQuery;
+    }
+
+    /**
+     * Get our rollup query
+     *
+     * @return Select
+     */
+    protected function rollupQuery()
+    {
+        if ($this->rollupQuery === null) {
+            $this->rollupQuery = $this->prepareRollupQuery();
+        }
+
+        return $this->rollupQuery;
+    }
+
+    /**
+     * Add a specific named dimension
+     *
+     * @param string $name
+     * @return $this
+     */
+    public function addDimensionByName($name)
+    {
+        $this->addDimension($this->createDimension($name));
+
+        return $this;
+    }
+
+    /**
+     * We first prepare the queries and to finalize it later on
+     *
+     * This way dimensions can be added one by one, they will be allowed to
+     * optionally join additional tables or apply other modifications late
+     * in the process
+     *
+     * @return void
+     */
+    protected function finalizeInnerQuery()
+    {
+        $innerQuery = $this->prepareInnerQuery();
+        $resolver = $innerQuery->getResolver();
+        $dimensions = $this->listDimensions();
+        $sourceTable = $innerQuery->getModel()->getTableName();
+
+        $groupBy = [];
+        foreach ($resolver->resolveRelations($sourceTable . '.vars') as $relation) {
```

Comment:
The entire relation resolve stuff has to be performed by addToCube. It's a *custom variable* dimension after all. The knowledge how to reach the custom variables from the base table must be part of it. The cube on the other side only knows it's some kind of *dimension*, but not which one exactly.